src/remote-editor-service.ts

```diff
@@ -0,0 +1,268 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as child_process from 'child_process';
+import * as util from 'util';
+
+import {SourceRange} from './ast/ast';
+import {BaseEditor, EditorService, Position, TypeaheadCompletion, Warning} from './editor-service';
+import {FSUrlLoader} from './url-loader/fs-url-loader';
+import {PackageUrlResolver} from './url-loader/package-url-resolver';
+
+interface RequestWrapper {
+  id: number;
+  value: Request;
+}
+interface ResponseWrapper {
+  id: number;
+  value: SettledValue;
+}
+type SettledValue = Resolution | Rejection;
+interface Resolution {
+  kind: 'resolution';
+  resolution: any;
+}
+interface Rejection {
+  kind: 'rejection';
+  rejection: string;
+}
+
+interface Deferred<V> {
+  resolve: (resp: V) => void;
+  reject: (err: any) => void;
+  promise: Promise<V>;
+}
+
+type Request = InitRequest | FileChangedRequest | GetWarningsRequest |
+    GetDocumentationRequest | GetDefinitionRequest |
+    GetTypeaheadCompletionsRequest | ClearCachesRequest;
+interface InitRequest {
+  kind: 'init';
+  basedir: string;
+}
+interface FileChangedRequest {
+  kind: 'fileChanged';
+  localPath: string;
+  contents?: string;
+}
+interface GetWarningsRequest {
+  kind: 'getWarningsFor';
+  localPath: string;
+}
+interface GetDocumentationRequest {
+  kind: 'getDocumentationFor';
+  localPath: string;
+  position: Position;
+}
+interface GetDefinitionRequest {
+  kind: 'getDefinitionFor';
+  localPath: string;
+  position: Position;
+}
+interface GetTypeaheadCompletionsRequest {
+  kind: 'getTypeaheadCompletionsFor';
+  localPath: string;
+  position: Position;
+}
+interface ClearCachesRequest {
+  kind: 'clearCaches';
+}
+
+/**
+ * Runs this file (remote-editor-service.js) in a new node process and
+ * exposes a promise based request API for communicating with it.
+ */
+class SelfChannel {
```

Comment:
Can you expand on that? What questions do you have that aren't answered by the comment on this class?